[Builtin][OSLog] Create a builtin and SPI for OSLog to get a pointer to the global string table section for strings constructed from string literals.

ravikandhadai opened the pull request:
This PR is split into three commits: 

- Create a named builtin: globalStringTablePointer: String -> Builtin.RawPointer that accepts a String and returns a pointer to a location in the global section if the string is constructed from a string literal, or traps otherwise. 

- Teach constant folding pass to fold the uses of the builtin to a `string_literal` SIL instruction when the builtin is used on a string constructed from a literal through the 'makeUTF8' string initializer. 

- Create a SPI in the stdlib so that the new OSLog APIs can use this builtin (via the SPI) instead of `withCString` to pass a pointer to the format string to the _os_log_impl library function, with the guarantee that the pointer will be pointing to the global section of the binary. 

<rdar://51722379>

ravikandhadai committed:
[Builtin][SILGen][IRGen] Create a new builtin "globalStringTablePointer":
String -> Builtin.RawPointer that given a string constructed from a
literal, returns the address of the string literal in the global string
table of the compiled binary as a pointer.

ravikandhadai commented:
@swift-ci Please test

milseman commented on test/IRGen/builtins.swift:
```diff
@@ -804,7 +804,15 @@ enum MyError : Error {
+// CHECK: ret i8* undef
+func globalStringTablePointerUse(_ str: String) -> Builtin.RawPointer {
+  return Builtin.globalStringTablePointer(str);
+}
```
Should it compile to a trap, or fail to compile?

ravikandhadai replied:
It compiles to a trap to enable them to wrapped inside other `@_transparent` functions,  and eventually some client function (possibly form a different module) define the constant string.

E.g. the SPI `_globalStringTablePointer` in `Builtin.swift` wraps this builtin and applies it to a non-constant argument. This SPI must compile as a part of stdlib.  Also, the SPI itself is used the `osLog` function which still applies the builtin to a non-constant string obtained from a parameter. Eventually, the constant is produced only when the client of `osLog` is seen.  Also each of these functions: `_globalStringTablePointer`, `osLog` are compiled into separate modules and so the calling context is not available when they are compiled.

ravikandhadai replied:
May be, we can emit a diagnostics if the builtin is applied to a non-literal string in a non-transparent function. However, we need the IRGen to emit a trap for transparent functions.

ravikandhadai replied:
Btw, based on my discussions,  if `@_alwaysEmitIntoClient` functions are IRGen'ed lazily like private methods of classes, it would be possible to make the IRGen for this builtin: `Bultin.globalStringTablePointer` as `llvm::unreachable`. However, as of now, `@_alwaysEmitIntoClient` will pass through IRGen whether or not they are used within the module.

ravikandhadai commented:
@swift-ci Please test Linux Platform

ravikandhadai commented:
@swift-ci Please smoke test Linux Platform

ravikandhadai commented:
@swift-ci Please test Linux Platform

atrick reviewed (approved):
Great, thanks!

ravikandhadai committed:
[Constant Folding][OSLogOptimization] Fold the uses of the builtin
"globalStringTablePointer": String -> Builtin.RawPointer` to a
string_literal instruction if the string that is passed is constructed
from a literal. Otherwise, emit diagnostics.

ravikandhadai committed:
[stdlib][oslog] Create an SPI _globalStringTablePointer that exposes
the builtin.globalStringTablePointer to the new OSLog overlay.

Modify the new OSLog implementation to use this SPI instead of
`withCString` to pass the (compiler-generated) format string to
the C os_log_impl ABI.

Move the OSLogOptimization pass before constant propagation in
the pass pipeline so that the SPI and the builtin it uses can be
folded to a string_literal instruction.

Update OSLogTests to work with the changes in the implementation.

ravikandhadai commented:
Pushed a commit addressing the comments

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

compnerd commented:
I believe that this broke the Windows SDK build.  https://dev.azure.com/compnerd/windows-swift/_build/results?buildId=3992&view=logs&j=b9e62f99-1a98-5ed7-01d2-f4794231ed79  Could you please take a look?

ravikandhadai commented:
@compnerd It looks like the new builtin introduced by this commit is not found by the bot when building the standard library. The commit changes the compiler to include this new builtin. So, the compiler should be rebuilt and it must be used to build the standard library. I guess that's happening? Perhaps you could also try a clean build.
